keybinding concept

danielsteinkogler opened the pull request:
<!--
Check relevant points but **please do not remove entries**.
-->

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [x] I added a line to [changelog.md](/doc/changelog.md)
- [x] The PR is rebased with current master.
- [x] Details of what you changed are in commit messages.
- [ ] References to issues, e.g. `close #X`, are in the commit messages and changelog.
- [x] The buildserver is happy.

<!--
If you have any troubles fulfilling these criteria, please write about the trouble as comment in the PR.
We will help you, but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For documentation fixes, spell checking, and similar none of these points below need to be checked.
Otherwise please check these points when getting a PR done:
-->

- [ ] I have installed and I am using [pre-commit hooks](../doc/contrib/README.md#Hooks)
- [ ] I fully described what my PR does in the documentation
- [ ] I fixed all affected documentation
- [ ] I fixed the introduction tour
- [ ] I wrote migrations in a way that they are compatible with already present data
- [ ] I fixed all affected decisions
- [ ] I added automated tests or a [manual test protocol](../doc/tests/manual/protocol.md)
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](https://github.com/ElektraInitiative/PermaplanT/blob/master/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)
- [ ] Exceptions to any guidelines are documented

## Review

<!--
Reviewers can copy&check the following to their review.
Also the checklist above can be used.
But also the PR creator should check these points when getting a PR done:
-->

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] I've read through the whole documentation
- [ ] I've checked conformity to guidelines

danielsteinkogler commented:
I added a new concept for our keybinding solution that comes a little bit closer to the solution @tschawax described in https://github.com/ElektraInitiative/PermaplanT/pull/944

markus2330 commented on doc/decisions/frontend_keyhandling.md:
```diff
@@ -0,0 +1,65 @@
+
+- **Ease of Integration**: The keybinding system should be straightforward to integrate into different components of the application.
+
+- **User Configuration**: We should plan for future functionality that allows users to configure keybindings according to their preferences.
```
This is not needed. It should only be developer-friendly: developers should easily get to know where which key binding is assigned.

Having all features to great good UX/UI is however needed (e.g. scoping, maybe @tschawax has more input on that).

markus2330 commented on doc/decisions/frontend_keyhandling.md:
```diff
@@ -0,0 +1,65 @@
+
+- **User Configuration**: We should plan for future functionality that allows users to configure keybindings according to their preferences.
+
+## Decision / Current Solution
```
Before Deciding, give solutions (see template)

markus2330 commented on doc/decisions/frontend_keyhandling.md:
````diff
@@ -0,0 +1,65 @@
+}
+```
+
+Components that require keybindings can define action handlers for the configured actions. Here's an example of how this can be implemented:
````
Please one sentence per line to easier comment on individual statements:

```suggestion
Components that require keybindings can define action handlers for the configured actions.
Here's an example of how this can be implemented:
```

markus2330 commented on doc/decisions/frontend_keyhandling.md:
```diff
@@ -0,0 +1,65 @@
+
+While it's easy to add key listeners to focused components like form inputs, it becomes more challenging when dealing with a component like the Konva map. This is because we cannot directly bind keys to the map layer.
+
+Furthermore there are instances where we want to handle keys for map feature even when another component is focused. For example, if a user selects a plant in the right toolbar, pressing the "Escape" key should probably also exit planting mode.
```
What we want from UI perspective should be described in @tschawax document (#918)

markus2330 commented on doc/decisions/frontend_keyhandling.md:
```diff
@@ -0,0 +1,65 @@
+Furthermore there are instances where we want to handle keys for map feature even when another component is focused. For example, if a user selects a plant in the right toolbar, pressing the "Escape" key should probably also exit planting mode.
+If we also want to clear a search with the "Escape" key, we have to stop event propagation.
+
+### Interesting npm package
+
+During research following npm package was found which seems to be very promising:
+
+- [react-hotkeys-hook](https://github.com/JohannesKlauss/react-hotkeys-hook/tree/main)
+
+It provides a hook for binding keys in components and also offers scoping abilities.
+
+### Ruled out alternative
+
+- [react-hotkeys](https://www.npmjs.com/package/react-hotkeys): provides a component that receives a keymap and actionhandlers similar to our planned concept and listens to events if a child of the component is in focus. It seams to be very suitable for encapsulating events, however it doesn't really solve the challenge of focus konva layers and although the librarie is widely used the last reala ...
```
Put both up to solutions. I agree we can rule out react-hotkeys, as scoping is important from UI perspective.

```suggestion
### React Hotkeys Hook

[react-hotkeys-hook](https://github.com/JohannesKlauss/react-hotkeys-hook/tree/main)

- It provides a hook for binding keys in components and also
- offers scoping abilities.

### React Hotkeys

 [react-hotkeys](https://www.npmjs.com/package/react-hotkeys): provides a component that receives a keymap and actionhandlers similar to our planned concept and listens to events if a child of the component is in focus.

- It seems to be very suitable for encapsulating events, however
- it doesn't really solve the challenge of focus konva layers and
- although the library is widely used the last release was 4 years ago.
```

markus2330 reviewed (changes requested):
As the concept is so central to PermaplanT and basically everything but backend is strongly effected, we can discuss it on our regular meeting on Monday.

- https://www.npmjs.com/package/react-hotkeys-hook should be definitely looked into, great point.
- Please also have a look what #1000 is trying to do.

danielsteinkogler commented:
I restructured the documentation a little bit. I moved everything to solution instead of decision because actually there is no final decision made yet. I look forward to discuss following questions on monday:

- Where to bind keylisteners (especially for konva layers)?
- Should we type available keys and actions (enums?, constants?)?
- Do we want to use https://www.npmjs.com/package/react-hotkeys-hook?

danielsteinkogler commented:
i updated the concept and moved all refactorings/prove of concept to a seperate PR (https://github.com/ElektraInitiative/PermaplanT/pull/1016)
this PR is currently a draft because i want to add some method documentation tomorrow

markus2330 commented on doc/decisions/frontend_keyhandling.md:
```diff
@@ -0,0 +1,108 @@
+- There should be a single source of truth
+- it should be easy to understand for developers.
+
+- **Where do we add key listeners?**
+
+- While it's easy to add key listeners to focused components like form inputs, it becomes more challenging when dealing with a component like the Konva map.
+- This is because we cannot directly bind keys to the map layer.
```
I wouldn't squeeze that in the same decision. It probably doesn't need a decision but just a way that works nicely. (Concept)

markus2330 commented on doc/decisions/frontend_keyhandling.md:
```diff
@@ -0,0 +1,108 @@
+- While it's easy to add key listeners to focused components like form inputs, it becomes more challenging when dealing with a component like the Konva map.
+- This is because we cannot directly bind keys to the map layer.
+
+#### Interesting npm package
```
Please create a subsubsection for each possible solution.

markus2330 commented on doc/decisions/frontend_keyhandling.md:
```diff
@@ -0,0 +1,108 @@
+
+### Keybinding Configuration
+
+A json file contains all keybindings that are used in the application and serves the following purpose
```
```suggestion
A JSON file contains all keybindings that are used in the application and serves the following purpose
```

markus2330 commented on doc/decisions/frontend_keyhandling.md:
```diff
@@ -0,0 +1,108 @@
+
+A json file contains all keybindings that are used in the application and serves the following purpose
+
+- keybindings can easily be changed by editing the file
```
Consistency within itemize

```suggestion
- Keybindings can easily be changed by editing the file.
```

markus2330 commented on doc/decisions/frontend_keyhandling.md:
```diff
@@ -0,0 +1,108 @@
+    "navigateToMap": "Ctrl+m"
+  }
+  "planting": {
+    "exitPlantingMode": "Escape"
+    "search":{
+      "clearSearch": "Escape"
```
Can we use some standard names? Toolkits usually provide such names.

```suggestion
    "exitPlantingMode": "Esc"
    "search":{
      "clearSearch": "Esc"
```

markus2330 commented on doc/decisions/frontend_keyhandling.md:
```diff
@@ -0,0 +1,108 @@
+- there can also be nested scopes
+- keys can be combined with '+'
+
+### Keyhandling on focused elements
```
```suggestion
### Keyhandling on Focused Elements
```

markus2330 commented on doc/decisions/frontend_keyhandling.md:
```diff
@@ -0,0 +1,108 @@
+### Keyhandling on focused elements
+
+- for components that provide keylisteners and are usually focused on use (e.g. input fields)
+- method: add scope to json, bind a keyhandler and execute action according to configuration
```
```suggestion
- method: add scope to JSON, bind a keyhandler and execute action according to configuration
```

markus2330 commented on doc/decisions/frontend_keyhandling.md:
```diff
@@ -0,0 +1,108 @@
+  "global": {
+    "navigateToMap": "Ctrl+m"
+  }
+  "planting": {
```
How can we distinguish layer names, function names etc. here?

markus2330 commented on doc/decisions/frontend_keyhandling.md:
````diff
@@ -0,0 +1,108 @@
+
+- for components that provide keylisteners and are usually focused on use (e.g. input fields)
+- method: add scope to json, bind a keyhandler and execute action according to configuration
+- if actions of parent components shouldn't be triggered, event propagation must be stopped
+
+Example
+
+```
+    const handleKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
+      const action = getActionNameFromKeyEvent("planting.search", event)
+      if(action === "clearSearch"){
+        resetSearch();
+        event.stopPropagation();
+      }
+    };
+```
````
move this concept to guidelines or solution chapter of architecture.

markus2330 reviewed (changes requested):
Thx for the improvements. Currrently it is still not a clear decision. For a decision several possibilities should be described, and the best chosen (which packages were not used and why etc.). For some parts of the "decision" it is probably better if you directly write a guideline/concept for how it should work.

markus2330 commented on doc/guidelines/frontend-keyhandling.md:
```diff
@@ -0,0 +1,80 @@
+  },
+};
+
+useKeyHandlers(createKeyHandlersFromConfig("planting_layer", keyHandlerActions));
```
`useKeyHandlers` is a function from you? Please introduce it here shortly or at least add comments in the example.

In #1016 only createKeyBindingsAccordingToConfig exists.

Please also make sure the snippets are formatted correctly.

danielsteinkogler replied:
useKeyHandlers is the custom hook we introduced in a earlier PR that is already merged. i will add a comment.

markus2330 commented on doc/decisions/frontend_keybinding.md:
Rename file to frontend_keybinding.md

markus2330 reviewed (changes requested):
Looks very good to me. :rocket: Only formatting/stylistic changes required. Terms are a bit inconsistent, but I think it should be easy to fix as they only seem to be left-overs.

absurd-turtle commented on doc/decisions/frontend_keybinding.md:
```diff
@@ -0,0 +1,38 @@
+- Link: [react-hotkeys-hook](https://github.com/JohannesKlauss/react-hotkeys-hook/tree/main)
+- provides a hook for binding keys in components and also
+- offers scoping abilities.
+- however, integration seems not to be as straightforward as with custom solution and so its doesn't really solve a problem for us.
```
```suggestion
- however, integration seems not to be as straightforward as with a custom solution and so it doesn't really solve a problem for us.
```

absurd-turtle commented on doc/decisions/frontend_keybinding.md:
```diff
@@ -0,0 +1,38 @@
+## Rational
+
+- We already have a custom keybinding hook and currently don't need any of the features provided by the packages.
+- the custom hook gives us more flexibility and is easier to integrate into our components.
```
```suggestion
- The custom hook gives us more flexibility and is easier to integrate into our components.
```

absurd-turtle commented on doc/guidelines/frontend-keyhandling.md:
````diff
@@ -0,0 +1,80 @@
+```
+{
+  "global": {
+    "<<action_name>>": ["<<Key-Binding>>, ...]"
````
I'd suggest swapping the key and the value here if it is not too much effort:
"<<key_binding>>": "<<action_name>>"
The reason is that I think it is easier to look up the corresponding action within the event handler: e.g. config.global['Escape'] to get the corresponding action.
It is also easier to check if there are keybindings assigned multiple times as the json is not valid anymore.

markus2330 replied:
I also suggested that already but everyone else found it more clear as it is proposed now. A check if key-bindings are used several times should be quite simple in both cases.

> config.global['Escape']

This looks to me to a reason to keep it as proposed. The code shouldn't bother about which keybinding is used, otherwise you couldn't simply make changes in the JSON. Or did I misunderstand?

absurd-turtle replied:
Fair enough. I don't have a strong opinion about it, it is just more intuitive for me as in my experience keybindings are often defined with binding -> action in my experience.
For example vim config, sxhkd config but that doesn't necessarily mean that it is better for us :)
It is not hard to check for duplicates in the config file but my point was that you have to do it.
When the keybinding is the key we have an invalid json as soon as the keybinding is used twice.
On the other hand it might be quicker to look up the bindings from the config with the action as key so we can leave it like that :)

absurd-turtle reviewed (changes requested):
Looks pretty good to me :)
I would change the structure of the keybinding config file but other than that everything sounds reasonable.

danielsteinkogler commented:
jenkins build please

danielsteinkogler commented:
jenkins build please

danielsteinkogler commented:
@markus2330 i think we can merge this after approval

markus2330 commented on doc/guidelines/frontend-keyhandling.md:
```diff
@@ -0,0 +1,80 @@
+- **Developer-friendly**: Developers should get an overview where which key binding is assigned.
+- **UX**: Collisions between shortcuts which trigger multiple actions at once should be avoided.
+
+### Important notes for Konva Key handling
```
There are several change requests still open and also rebase is required.

danielsteinkogler replied:
oh sorry for some reason i didn't notice that there a further change requests. i will take care of them now

danielsteinkogler commented on doc/guidelines/frontend-keyhandling.md:
```diff
@@ -0,0 +1,80 @@
+
+### Important notes for Konva Key handling
+
+- since keys cannot be bound directly on konva elements, keyhandlers have to check if the corresponding layer is active to avoid collisions
```
i added a checkbox/todo here: https://github.com/ElektraInitiative/PermaplanT/issues/1031.
For implementing this feature I need to check if a layer is active. After implementing this, I will provide specific guidelines on how to do it

markus2330 replied:
Ok, the macOS thing, if it is not trivial/obvious, can also be done with #1031.

danielsteinkogler commented on doc/guidelines/frontend-keyhandling.md:
```diff
@@ -0,0 +1,80 @@
+- there are different sections for keybindings:
+  - global: keybindings that are globally active
+  - scopes for specific layers: keybindings that are only active if the corresponding layer is active. this scopes should be named according to the layer name
+  - _just_for_documentation_: keybindings that are only used for documentation purposes and are not meant to get configured
```
i added an example to make clear what i meant

danielsteinkogler commented:
@markus2330 i rebased the pr and resolved the pr comments. it still says that the requested changes have to be addressed. However, i don't see any unresolved comments. Is there a view where i can see only unresolved comments in github?

markus2330 commented:
> @markus2330 i rebased the pr and resolved the pr comments. it still says that the requested changes have to be addressed.
> Or do you have to approve the changes?

Yes, the normal way is that you rerequest reviews when everything is done. :rocket: 

> However, i don't see any unresolved comments. Is there a view where i can see only unresolved comments in github?

I personally use E-Mail notification and go through them one by one.

https://github.com/ElektraInitiative/PermaplanT/pull/994/files might sometimes be helpful,  it shows unresolved comments, but only those who refer to lines that exist. We can discuss such workflows on Monday.

You self-conflicted in the doc/changelog.md. Try a better random generator :wink:

markus2330 commented on doc/guidelines/frontend-keyhandling.md:
```diff
@@ -0,0 +1,90 @@
+
+### Keybinding Configuration
+
+The file 'keybindings.json' contains all keybindings that are used in the application and serves the following characteristics:
```
```suggestion
The file `config/keybindings/keybindings.json` contains all keybindings that are used in the application and serves the following characteristics:
```

Please focus a bit more

markus2330 commented on doc/guidelines/frontend-keyhandling.md:
```diff
@@ -0,0 +1,90 @@
+
+The file 'keybindings.json' contains all keybindings that are used in the application and serves the following characteristics:
+
+- keybindings can easily be changed by editing the file
+- developers have an overview of all keybindings
```
These are not really guidelines?

```suggestion
```

markus2330 commented on doc/guidelines/frontend-keyhandling.md:
```diff
@@ -0,0 +1,90 @@
+- keybindings can easily be changed by editing the file
+- developers have an overview of all keybindings
+- there are different sections for keybindings:
+  - global: keybindings that are globally active
```
(Also missing at other places)

```suggestion
  - `global`: keybindings that are globally active
```

markus2330 commented on doc/guidelines/frontend-keyhandling.md:
```diff
@@ -0,0 +1,90 @@
+    Keybindings that are only used for documentation purposes since the keybinding is hardcoded in the component.
+    For example, the search field can be cleared by pressing the Escape key, which is typically a native browser action and should not be configurable in the keybindings file.
+- modifier keys are supported (Shift, Alt, Ctrl, Meta)
+  - they can be used in combination with other keys by using the '+' sign (e.g. 'Alt+P', 'Shift+A')
```
How do these map on Mac OS?
Can you maybe introduce a wildcard modifier key so that it automatically works on Mac OS, too?

danielsteinkogler replied:
the react key event supports the four mentioned modifier keys (Shift, Alt, Ctrl, Meta). 
The keys should be mapped like this:

- event.ctrlKey: Represents whether the Control key (Ctrl) is pressed.
- event.shiftKey: Represents whether the Shift key is pressed.
- event.altKey: Represents whether the Alt key is pressed (Option key on macOS).
- event.metaKey: Represents whether the Meta key (Command key on macOS) is pressed.

Unfortunately, i cannot test if it is really working like this. I created a left over issue: https://github.com/ElektraInitiative/PermaplanT/issues/1057

markus2330 commented on doc/guidelines/frontend-keyhandling.md:
```diff
@@ -0,0 +1,90 @@
+### Keybinding Configuration
+
+The file 'keybindings.json' contains all keybindings that are used in the application and serves the following characteristics:
+
```
```suggestion

- for all features add commonly-used and vim-like bindings
- we generally try to provide several keybindings for the same functionality, to make the functionality easier accessible
```

markus2330 commented on doc/guidelines/frontend-keyhandling.md:
````diff
@@ -0,0 +1,90 @@
+}
+```
+
+### Keyhandling on Focused Elements
````
First describe what should be done in any case. Ideally add a step-by-step description what someone should do to properly add a keybinding. Then add info about special cases.

danielsteinkogler replied:
in my opinion the subsections are not describing special cases. there are just two different ways we have to handle keyhandling for either focused elements or keybindings that should be available independent of focus. 
i created a more detailed step-by-step description for both variants.

markus2330 reviewed (changes requested):
Much better, a bit more rework is still needed though.

markus2330 commented:
I removed "please merge" as I added comments, please readd when you are done (left-over tasks simply to an left-over issue).
